File: CookieCuteStack.cs
using System;

/*
    CookieDataStructure
    Copyright (C) 2026 Pe4enbka008 (Helen Ivanova) 

    This code is free software: you can redistribute it and/or modify 
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or any later version. 

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

    See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/>.
*/


namespace smth
{
    /// <summary>
    /// Stack created and better-ed by Cookie :]
    /// </summary>
    /// <typeparam name="T">Type of the stack</typeparam>
    public class CookieStack<T>
    {
        private CookieNode<T>? nodes;

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack() { this.nodes = null; }

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack(T value) { this.nodes = new CookieNode<T>(value); }

        /// <summary>
        /// Class constructor  -- not bagrut supported!
        /// </summary>
        public CookieStack(CookieNodeList<T> list) { this.nodes = CookieNodeWorker.NodeListToNodes(list); }

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack(CookieNode<T> nodes) { this.nodes = nodes; }

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack(T[] list) { this.nodes = CookieNodeWorker.ArrayToNodes(list); }


        /// <summary>
        /// Returns if the stack is empty
        /// </summary>
        /// <returns>true is the stack is empty</returns>
        public bool IsEmpty() { return this.nodes == null; }

        /// <summary>
        /// Returns stack's length  -- not bagrut supported!
        /// </summary>
        public int Length { get { return CookieNodeWorker.RecursionCount<T>(this.nodes); } }


        // Getters-Setters
        /// <summary>
        /// LIFO (LAST IN ; FIRST OUT) - puts the value in front
        /// </summary>
        /// <param name="value">Value to save</param>
        public void Push(T value) 
        { this.nodes = new CookieNode<T>(value, this.nodes); }


        /// <summary>
        /// FILO (FIRST IN ; LAST OUT) - gets the front value ; could be null
        /// </summary>
        /// <returns>The top (last added) value</returns>
        public T? PopValue()
        {
            if (this.nodes == null)
                return default;  // cause.... Just cause >:]

            CookieNode<T> node = this.nodes;
            this.nodes = node.GetNext();
            node.Next = null;
            return node.Value;
        } // Pop

        /// <summary>
        /// Easier way of Pop function :]  -- not bagrut supported!
        /// </summary>
        public T? Pop { get { return this.PopValue(); } }


        /// <summary>
        /// Gets the top value, if the stack is empty, returns default of the type
        /// </summary>
        /// <returns>value of the top</returns>
        public T? GetTop()
        { return this.nodes != null ? this.nodes.Value : default; } // if (this.nodes != null) return this.nodes.Value; return default; 


        /// <summary>
        /// Creates a copy of the object in type of the Nodes  -- not bagrut supported!
        /// </summary>
        /// <returns>copy of the node list</returns>
        public CookieNode<T>? Copy()
        {
            if (this.nodes == null)
                return null;

            CookieNode<T> return_value = new CookieNode<T>(this.nodes.Value);
            CookieNode<T> current = return_value;
            CookieNode<T>? nodes = this.nodes.Next;

            while (nodes != null)
            {
                current.Next = new(nodes.Value);
                current = current.Next;
                nodes = nodes.Next;
            } // while

            return return_value;
        } // Copy


        /// <summary>
        /// Clears the Stack
        /// </summary>
        public void Clear()
        { this.nodes = null; }



        // override
        /// <summary>
        /// override for ToString to - ['value', 'value', 'value', ...]
        /// </summary>
        /// <returns>string of the class</returns>
        public override string ToString()
        {
            if (this.nodes == null)
                return "[]";

            string str = "[";
            CookieNode<T>? node = this.nodes;

            while (node != null)
            {
                str += node.ToString();
                if (node.GetNext() != null)
                    str += ", ";
                node = node.GetNext();
            } // while

            return str + "]";
        } // override ToString

        /// <summary>
        /// override for ToString to - ['value'{split} 'value'{split} 'value'{split} ...]
        /// </summary>
        /// <returns>string of the class</returns>
        public string ToString(string split)
        {
            if (this.nodes == null)
                return "[]";

            string str = "[";
            CookieNode<T>? node = this.nodes;

            while (node != null)
            {
                str += node.ToString();
                if (node.GetNext() != null)
                    str += split;
                node = node.GetNext();
            } // while

            return str + "]";
        } // override ToString

    } // class CookieStack


    /// <summary>
    /// Queue created and better-ed by Cookie :]
    /// </summary>
    /// <typeparam name="T">Type of the queue</typeparam>
    public class CookieQueue<T>
    {
        private CookieNode<T>? nodes;
        private CookieNode<T>? pointer_to_end;


        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieQueue() { this.nodes = null; this.pointer_to_end = null; }

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieQueue(T value) 
        { this.nodes = new CookieNode<T>(value); this.pointer_to_end = this.nodes; }


        /// <summary>
        /// Class constructor  -- not bagrut supported!
        /// </summary>
        public CookieQueue(CookieNodeList<T> list) 
        { 
            this.nodes = CookieNodeWorker.NodeListToNodes(list);
            this.pointer_to_end = this.nodes;
            while (this.pointer_to_end.Next != null)
                this.pointer_to_end = this.pointer_to_end.Next; // got to the last
        } // __init__


        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieQueue(CookieNode<T> nodes) 
        { 
            this.nodes = nodes;
            this.pointer_to_end = this.nodes;
            while (this.pointer_to_end.Next != null)
                this.pointer_to_end = this.pointer_to_end.Next; // got to the last
        } // __init__

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieQueue(T[] list) 
        { 
            this.nodes = CookieNodeWorker.ArrayToNodes(list);
            this.pointer_to_end = this.nodes;
            while (this.pointer_to_end.Next != null)
                this.pointer_to_end = this.pointer_to_end.Next; // got to the last
        } // __init__


        /// <summary>
        /// Returns if the queue is empty
        /// </summary>
        /// <returns>true is the queue is empty</returns>
        public bool IsEmpty() 
        { return this.nodes == null && this.pointer_to_end == null; }

        /// <summary>
        /// Returns queue's length  -- not bagrut supported!
        /// </summary>
        public int Length { get { return CookieNodeWorker.RecursionCount<T>(this.nodes); } }


        // Getters-Setters
        /// <summary>
        /// FIFO (FIRST IN ; FIRST OUT) - puts the value at the end
        /// </summary>
        /// <param name="value">Value to save</param>
        public void Insert(T value)
        { 
            if (this.nodes == null) // queue is empty
            {
                this.nodes = new CookieNode<T>(value);
                this.pointer_to_end = this.nodes;
                return;
            } // if 
            this.pointer_to_end.SetNext(new CookieNode<T>(value));
            this.pointer_to_end = this.pointer_to_end.Next;
        } // Insert


        /// <summary>
        /// FIFO (FIRST IN ; FIRST OUT) - gets the front value ; could be null
        /// </summary>
        /// <returns>The top (first added) value</returns>
        public T? RemoveValue()
        {
            if (this.nodes == null)
                return default;  // cause.... Just cause >:]

            CookieNode<T> node = this.nodes;
            if (node.GetNext() == null)
                this.pointer_to_end = null;
            this.nodes = node.GetNext();
            node.Next = null;
            return node.Value;
        } // Remove

        /// <summary>
        /// Easier way of Remove function :]  -- not bagrut supported!
        /// </summary>
        public T? Remove { get { return this.RemoveValue(); } }


        /// <summary>
        /// Gets the top value, if the stack is empty, returns default of the type
        /// </summary>
        /// <returns>value of the top</returns>
        public T? GetTop()
        { return this.nodes != null ? this.nodes.Value : default; } // if (this.nodes != null) return this.nodes.Value; return default; 


        /// <summary>
        /// Creates a copy of the object in type of the Nodes  -- not bagrut supported!
        /// </summary>
        /// <returns>copy of the node list</returns>
        public CookieNode<T>? Copy()
        {
            if (this.nodes == null)
                return null;

            CookieNode<T> return_value = new CookieNode<T>(this.nodes.Value);
            CookieNode<T> current = return_value;
            CookieNode<T>? nodes = this.nodes.Next;

            while (nodes != null)
            {
                current.Next = new(nodes.Value);
                current = current.Next;
                nodes = nodes.Next;
            } // while

            return return_value;
        } // Copy


        /// <summary>
        /// Clears the Queue
        /// </summary>
        public void Clear()
        { this.nodes = null; this.pointer_to_end = null; }



        // override
        /// <summary>
        /// override for ToString to - ['value', 'value', 'value', ...]
        /// </summary>
        /// <returns>string of the class</returns>
        public override string ToString()
        {
            if (this.nodes == null)
                return "[]";

            string str = "[";
            CookieNode<T>? node = this.nodes;

            while (node != null)
            {
                str += node.ToString();
                if (node.GetNext() != null)
                    str += ", ";
                node = node.GetNext();
            } // while

            return str + "]";
        } // override ToString

        /// <summary>
        /// override for ToString to - ['value'{split} 'value'{split} 'value'{split} ...]
        /// </summary>
        /// <returns>string of the class</returns>
        public string ToString(string split)
        {
            if (this.nodes == null)
                return "[]";

            string str = "[";
            CookieNode<T>? node = this.nodes;

            while (node != null)
            {
                str += node.ToString();
                if (node.GetNext() != null)
                    str += split;
                node = node.GetNext();
            } // while

            return str + "]";
        } // override ToString

    } // class CookieQueue


} // namespace smth

File: Python/CookieCuteStack.py
"""
    CookieDataStructure
    Copyright (C) 2026 Pe4enbka008 (Helen Ivanova)

    This code is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

    See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/>.
"""
from CookieNode import CookieNode


class CookieStack:
    def __init__(self, value=None):
        """Constructor for CookieStack
        :param value: start nodes
        :type value: CookieNode | CookieStack | Any
        :return: nothing
        :rtype: None"""
        self.nodes = None
        if isinstance(value, CookieNode):
            self.nodes = value
        elif isinstance(value, CookieStack):
            for node in value.nodes:
                self.push(node.get_value())
        elif value is not None:
            self.nodes = CookieNode(value)

    def __str__(self):
        """Returns string value of the nodes
        :return: string value of the nodes
        :rtype: str"""
        return self.to_str(', ')

    def to_str(self, spliter):
        """Returns string value of the nodes
        :param spliter: spliter to put in between of the values
        :type spliter: str
        :return: string value of the nodes
        :rtype: str"""
        if not self.nodes:
            return '[]'

        return_value = '['
        walk_value = self.nodes
        while walk_value:
            return_value += str(walk_value)
            walk_value = walk_value.get_next()
            if walk_value:
                return_value += spliter

        return return_value + ']'

    def is_empty(self):
        """Returns true if the stack is empty
        :return: true if the stack is empty
        :rtype: bool"""
        return not self.nodes

    # getters-setters:

    def push(self, value):
        """FILO (FIRST IN ; LAST OUT) - puts the value in front
        :param value: value to push
        :type value: CookieNode | CookieStack | Any
        :return: nothing
        :rtype: None"""
        if isinstance(value, CookieNode):
            value.next_node = self.nodes
            self.nodes = value
        elif isinstance(value, CookieStack):
            for node in value.nodes:
                self.push(node.get_value())
        else:
            self.nodes = CookieNode(value, self.nodes)

    def pop(self):
        """LIFO (LAST IN ; FIRST OUT) - gets the front value ; could be null
        :return: value to pop
        :rtype: Any"""
        if self.is_empty():
            return None

        node = self.nodes
        self.nodes = self.nodes.next_node
        node.next_node = None
        return node.value

    def get_top(self):
        """Returns the top value of the stack
        :return: the top value of the stack
        :rtype: Any"""
        return None if self.nodes is None else self.nodes.value

    def copy(self):
        """Returns a copy of the stack
        :return: the copy of the stack
        :rtype: CookieStack"""
        if not self.nodes:
            return None

        return_value = CookieNode(self.nodes.value)
        current = return_value
        walk_value = self.nodes.next_node

        while walk_value:
            current.set_next(CookieNode(walk_value.value))
            current = current.get_next()
            walk_value = walk_value.get_next()

        return CookieStack(return_value)

    def clear(self):
        """Clears the stack
        :return: nothing
        :rtype: None"""
        self.nodes = None

    # length:
    def __len__(self):
        """Returns length of the stack
        :return: length
        :rtype: int"""
        if self.is_empty():
            return 0
        return sum(1 for _ in self.nodes)

    # reverse:
    def __reversed__(self):
        """reverses a stack
        :return: reversed stack
        :rtype: CookieStack"""
        if self is None:
            return None
        if len(self) <= 1:
            return self.copy()

        current = self.nodes
        rev = CookieStack()
        while current:
            rev.push(current.get_value())
            current = current.get_next()
        return rev

    # end


class CookieQueue:
    def __init__(self, value=None):
        """Constructor for CookieQueue
        :param value: start nodes
        :type value: CookieNode | CookieQueue | Any
        :return: nothing
        :rtype: None"""
        self.nodes = None
        self.end = None
        if isinstance(value, CookieNode):
            self.nodes = value
            self.end = self.nodes
        elif isinstance(value, CookieQueue):
            for node in value.nodes:
                new_node = CookieNode(node.get_value())
                self.insert(new_node)
                self.end = new_node
        elif value is not None:
            self.nodes = CookieNode(value)
            self.end = self.nodes

    def __str__(self):
        """Returns string value of the nodes
        :return: string value of the nodes
        :rtype: str"""
        return self.to_str(', ')

    def to_str(self, spliter):
        """Returns string value of the nodes
        :param spliter: spliter to put in between of the values
        :type spliter: str
        :return: string value of the nodes
        :rtype: str"""
        if not self.nodes:
            return '[]'

        return_value = '['
        walk_value = self.nodes
        while walk_value:
            return_value += str(walk_value)
            walk_value = walk_value.get_next()
            if walk_value:
                return_value += spliter

        return return_value + ']'

    def is_empty(self):
        """Returns true if the stack is empty
        :return: true if the stack is empty
        :rtype: bool"""
        return not self.nodes and not self.end

    # getters-setters:

    def insert(self, value):
        """FIFO (FIRST IN ; FIRST OUT) - puts the value in front
        :param value: value to push
        :type value: CookieNode | CookieQueue | Any
        :return: nothing
        :rtype: None"""
        # inserting another queue
        if isinstance(value, CookieQueue):
            current = value.nodes

            while current is not None:
                self.insert(current.value)  # fun recursion
                current = current.next_node
            return

        new_node = value if isinstance(value, CookieNode) else CookieNode(value)

        if self.is_empty():
            self.nodes = new_node
            self.end = new_node
            return

        self.end.next_node = new_node
        self.end = new_node

    def remove(self):
        """LFIFO (FIRST IN ; FIRST OUT) - gets the front value ; could be null
        :return: value to pop
        :rtype: Any"""
        if self.is_empty():
            return None

        node = self.nodes
        if node.next_node is None:
            self.end = None
        self.nodes = self.nodes.next_node
        node.next_node = None
        return node.value

    def get_top(self):
        """Returns the top value of the stack
        :return: the top value of the stack
        :rtype: Any"""
        return None if self.nodes is None else self.nodes.value

    def copy(self):
        """Returns a copy of the stack
        :return: the copy of the stack
        :rtype: CookieQueue"""
        if not self.nodes:
            return None

        return_value = CookieNode(self.nodes.value)
        current = return_value
        walk_value = self.nodes.next_node

        while walk_value:
            current.set_next(CookieNode(walk_value.value))
            current = current.get_next()
            walk_value = walk_value.get_next()
        return CookieQueue(return_value)

    def clear(self):
        """Clears the stack
        :return: nothing
        :rtype: None"""
        self.nodes = None
        self.end = None

    # length:
    def __len__(self):
        """Returns length of the queue
        :return: length
        :rtype: int"""
        if self.is_empty():
            return 0
        return sum(1 for _ in self.nodes)

    # reverse:
    def __reversed__(self):
        """reverses a queue
        :return: reversed queue
        :rtype: CookieQueue"""
        if self is None:
            return None
        if len(self) <= 1:
            return self.copy()

        current = self.nodes
        rev_stack = CookieStack()
        while current:
            rev_stack.push(current.get_value())
            current = current.get_next()

        rev = CookieQueue()
        while not rev_stack.is_empty():
            rev.insert(rev_stack.pop())
        return rev

    # end

